File: 프로그래머스/lv3/64062. 징검다리 건너기/징검다리 건너기.java
import java.util.*;

class Solution {
    
    private static class Stone{
        int idx, val;
        public Stone(int idx, int val){
            this.idx = idx;
            this.val = val;
        }
    }
    
    public int solution(int[] stones, int k) {
        
        PriorityQueue<Stone> pq = new PriorityQueue<>((o1, o2)->Integer.compare(o2.val, o1.val));
        
        for(int i = 0; i < k; i++){
            pq.add(new Stone(i, stones[i]));
        }       
        int answer = pq.peek().val;
        int now = k;
        while(now < stones.length){
            pq.add(new Stone(now, stones[now]));
            now++;
            while(pq.peek().idx < now - k){
                pq.poll();
            }
            answer = Math.min(answer, pq.peek().val);
        }
        return answer;
    }
}

File: 프로그래머스/lv3/64062. 징검다리 건너기/징검다리 건너기.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

int solution(vector<int> stones, int k) {
    
    priority_queue<pair<int, int>> pq;
    int i;
    for(i = 0; i < k; i++){
        pq.push({stones[i], i});
    }
    
    int answer = pq.top().first;
    while(i < stones.size()){
        pq.push({stones[i], i});
        i++;
        while(pq.top().second < i - k){
            pq.pop();
        }
        answer = min(answer, pq.top().first);
    }
    
    return answer;
}